### PR TITLE
DDF-5360 Adds the ability to go through OAuth to download a resource

### PR DESCRIPTION
**Pull request by blen-desta**
Needs: https://github.com/codice/ddf-ui/pull/83

#### What does this PR do?
Adds the ability to go through OAuth to download a resource.

#### Who is reviewing it? 
@SmithJosh @bakejeyner 

#### Select relevant component teams: 
@codice/security 

#### Ask 2 committers to review/merge the PR and tag them here.
@garrettfreibott
@stustison

#### How should this be tested?
- Verify downloading a resource from an OAuth protected source works. Download using
  - the download button
  - the inspector tab
  - the action tab
- Delete the user's tokens and try downloading a resource (should be prompted to log in).
- Verify querying with OAuth still works

#### What are the relevant tickets?
#5360 

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Threat Dragon models
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: platform/security/rest/security-rest-clientapi/src/main/java/org/codice/ddf/cxf/client/SecureCxfClientFactory.java [hunk 1]
```diff
@@ -42,6 +42,13 @@
    */
   WebClient getWebClient();
 
+  /**
+   * Returns the WebClient
+   *
+   * @return
+   */
+  WebClient getWebSystemClient();
```
Comment (stustison) [inline platform/security/rest/security-rest-clientapi/src/main/java/org/codice/ddf/cxf/client/SecureCxfClientFactory.java]: ❓ Why did you need to add this? It looks like it isn't used anywhere.

Comment (blen-desta) [reply platform/security/rest/security-rest-clientapi/src/main/java/org/codice/ddf/cxf/client/SecureCxfClientFactory.java]: It will not be used in DDF but downstream distributions will use it.

File: catalog/core/catalog-core-urlresourcereader/pom.xml [hunk 2]
```diff
@@ -154,12 +154,12 @@
                                         <limit implementation="org.codice.jacoco.LenientLimit">
                                             <counter>INSTRUCTION</counter>
                                             <value>COVEREDRATIO</value>
-                                            <minimum>0.70</minimum>
+                                            <minimum>0.68</minimum>
                                         </limit>
                                         <limit implementation="org.codice.jacoco.LenientLimit">
                                             <counter>BRANCH</counter>
                                             <value>COVEREDRATIO</value>
-                                            <minimum>0.64</minimum>
```
Comment (SmithJosh) [inline catalog/core/catalog-core-urlresourcereader/pom.xml]: ❓ Can you add a test for this?

Comment (blen-desta) [reply catalog/core/catalog-core-urlresourcereader/pom.xml]: Yes I will that whole method was not tested

File: catalog/core/catalog-core-urlresourcereader/src/main/java/ddf/catalog/resource/impl/URLResourceReader.java [hunk 3]
```diff
@@ -583,7 +594,8 @@ private boolean validateFilePath(File resourceFilePath) throws IOException {
     return false;
   }
 
-  protected WebClient getWebClient(String uri, Map<String, Serializable> properties) {
+  @VisibleForTesting
+  WebClient getWebClient(String uri, Map<String, Serializable> properties) {
```
Comment (SmithJosh) [inline catalog/core/catalog-core-urlresourcereader/src/main/java/ddf/catalog/resource/impl/URLResourceReader.java]: I get a compilation error when building. This test needs to be updated with the new `getWebClient` method signature: 
https://github.com/codice/ddf/blob/master/catalog/spatial/ogc/spatial-ogc-urlresourcereader/src/test/java/org/codice/ddf/spatial/ogc/catalog/resource/impl/ResourceReaderTest.java#L94-L95

**Comment by blen-desta** [comment]
build now

**Comment by blen-desta** [comment]
build now

**Comment by SmithJosh** [comment]
Successfully hero'd! Verified the following, first with DDF set up to federate the system token and then with it set up to federate the user's token:
* OAuth source query succeeds
* User can download the query results by clicking the `resource-download-url` in the Inspector.

File: platform/security/rest/security-rest-cxfwrapper/src/main/java/org/codice/ddf/cxf/oauth/OAuthSecurityImpl.java [hunk 4]
```diff
@@ -324,6 +326,7 @@ private String getNewAccessToken(
 
     Form formParam = new Form(GRANT_TYPE, grantType);
     queryParameters.forEach(formParam::param);
+    formParam.param(SCOPE, OPENID_SCOPE);
```
Comment (SmithJosh) [inline platform/security/rest/security-rest-cxfwrapper/src/main/java/org/codice/ddf/cxf/oauth/OAuthSecurityImpl.java]: I already approved, but we have a configuration for OAuth handler which determines the scopes that will be requested from the auth server. Wondering if we should include the full set from that config here as well

Comment (blen-desta) [reply platform/security/rest/security-rest-cxfwrapper/src/main/java/org/codice/ddf/cxf/oauth/OAuthSecurityImpl.java]: The problem is those that configuration is separate. This code is for federating and that is for login so they could be connected to two different providers.

**Comment by blen-desta** [comment]
build now

File: platform/security/rest/security-rest-cxfwrapper/src/main/java/org/codice/ddf/cxf/oauth/OAuthSecurityImpl.java (hunk 4 above)
Comment (bakejeyner) [reply platform/security/rest/security-rest-cxfwrapper/src/main/java/org/codice/ddf/cxf/oauth/OAuthSecurityImpl.java]: Do we see a use case where these federation scopes would need to be configurable? If so, I would imagine that would live in the source configuration somewhere, but 🤷‍♂

Comment (SmithJosh) [reply platform/security/rest/security-rest-cxfwrapper/src/main/java/org/codice/ddf/cxf/oauth/OAuthSecurityImpl.java]: Actually, it probably is fine as-is. It's likely we'll hit a situation where additional scopes are needed, but auth servers should be configurable to provide scopes as needed to specific clients without the client requesting them.

Comment (blen-desta) [reply platform/security/rest/security-rest-cxfwrapper/src/main/java/org/codice/ddf/cxf/oauth/OAuthSecurityImpl.java]: If that is the case we will need an additional configuration in the source configurations to take care of that.

Comment (SmithJosh) [reply platform/security/rest/security-rest-cxfwrapper/src/main/java/org/codice/ddf/cxf/oauth/OAuthSecurityImpl.java]: I was trying to say we shouldn't need to do that 😄 It makes sense to request the "openid" scope since that's required for OIDC, but for application-specific scopes, it should be possible to configure everything in the auth server rather than DDF. That's at least the case for Keycloak. You can provide a list of default scopes to issue on a per-client basis.

File: catalog/core/catalog-core-api/src/main/java/ddf/catalog/plugin/OAuthPluginException.java [hunk 5]
```diff
@@ -37,17 +46,38 @@ public int getStatusCode() {
     }
   }
 
-  public OAuthPluginException(String sourceId, String url, ErrorType errorType) {
```
[the remark this answers is not in the document]
Comment (blen-desta) [reply catalog/core/catalog-core-api/src/main/java/ddf/catalog/plugin/OAuthPluginException.java]: Yes

File: catalog/rest/catalog-rest-service/src/main/java/org/codice/ddf/rest/service/AbstractCatalogService.java [hunk 6]
```diff
@@ -372,6 +377,15 @@ public BinaryContent getDocument(
         // We need to catch this exception here or else execution will return to CXF and
```
Comment (SmithJosh) [inline catalog/rest/catalog-rest-service/src/main/java/org/codice/ddf/rest/service/AbstractCatalogService.java]: ✏️ Does this comment need to move down to the `catch (RuntimeException | UnsupportedEncodingException e)` line?

**Comment by blen-desta** [comment]
build now

**Comment by bakejeyner** [comment]
Hero ✔️ 

When I have an item in the inspector and delete the tokens off of the file system, I get a 500 when trying to download the resource. Re-searching fixes this issue.

**Comment by blen-desta** [comment]
build now
